#### 1. Import Data

In [1]:
import pandas as pd
%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
df = pd.read_csv('onlinepurchase.csv')

#### 2. Analyze Data

In [2]:
df.shape

(1022664, 9)

In [3]:
df.columns

Index(['Invoice', 'StockCode', 'Description', 'Quantity', 'Invoice_time',
       'Price', 'CustomerID', 'Country', 'Purchase_dt'],
      dtype='object')

In [4]:
df['CustomerID'].nunique()

5887

In [5]:
df['StockCode'].nunique()

5288

In [6]:
df['Purchase_dt'].unique()

array(['2010-01-04', '2010-01-05', '2010-01-06', '2010-01-07',
       '2010-01-08', '2010-01-10', '2010-01-11', '2010-01-12',
       '2010-01-13', '2010-01-14', '2010-01-15', '2010-01-17',
       '2010-01-18', '2010-01-19', '2010-01-20', '2010-01-21',
       '2010-01-22', '2010-01-24', '2010-01-25', '2010-01-26',
       '2010-01-27', '2010-01-28', '2010-01-29', '2010-01-31',
       '2010-02-01', '2010-02-02', '2010-02-03', '2010-02-04',
       '2010-02-05', '2010-02-07', '2010-02-08', '2010-02-09',
       '2010-02-10', '2010-02-11', '2010-02-12', '2010-02-14',
       '2010-02-15', '2010-02-16', '2010-02-17', '2010-02-18',
       '2010-02-19', '2010-02-21', '2010-02-22', '2010-02-23',
       '2010-02-24', '2010-02-25', '2010-02-26', '2010-02-28',
       '2010-03-01', '2010-03-02', '2010-03-03', '2010-03-04',
       '2010-03-05', '2010-03-07', '2010-03-08', '2010-03-09',
       '2010-03-10', '2010-03-11', '2010-03-12', '2010-03-14',
       '2010-03-15', '2010-03-16', '2010-03-17', '2010-

In [7]:
df.duplicated().sum()

11585

In [8]:
df=df.drop_duplicates()
df.shape

(1011079, 9)

In [9]:
df.head(20)

Invoice StockCode                          Description  Quantity  \
0    493410   TEST001              This is a test product.         5   
1   C493411     21539              RETRO SPOTS BUTTER DISH        -1   
2    493412   TEST001              This is a test product.         5   
3    493413     21724      PANDA AND BUNNIES STICKER SHEET         1   
4    493413     84578       ELEPHANT TOY WITH BLUE T-SHIRT         1   
5    493413     21723        ALPHABET HEARTS STICKER SHEET         1   
6    493414     21844                       RETRO SPOT MUG        36   
7    493414     21533            RETRO SPOT LARGE MILK JUG        12   
8    493414     37508      NEW ENGLAND CERAMIC CAKE SERVER         2   
9    493414    35001G                 HAND OPEN SHAPE GOLD         2   
10   493414     21527       RETRO SPOT TRADITIONAL TEAPOT         12   
11   493414     21531            RETRO SPOT SUGAR JAM BOWL        24   
12  C493415     21527       RETRO SPOT TRADITIONAL TEAPOT         -3   
13  C493426     22109         FULL ENGLISH BREAKFAST PLATE        -1   
14   493427     82483   WOOD 2 DRAWER CABINET WHITE FINISH         4   
15   493427     21681     GIANT MEDINA STAMPED METAL BOWL          2   
16   493427     21682     LARGE MEDINA STAMPED METAL BOWL          4   
17   493427    18096C  WHITE ROUND PORCELAIN TLIGHT HOLDER         6   
18   493427     21577      SAVE THE PLANET COTTON TOTE BAG         6   
19   493427     21576     LETS GO SHOPPING COTTON TOTE BAG         6   

        Invoice_time  Price  CustomerID         Country Purchase_dt  
0    2010-01-04 9:24   4.50     12346.0  United Kingdom  2010-01-04  
1    2010-01-04 9:43   4.25     14590.0  United Kingdom  2010-01-04  
2    2010-01-04 9:53   4.50     12346.0  United Kingdom  2010-01-04  
3    2010-01-04 9:54   0.85         NaN  United Kingdom  2010-01-04  
4    2010-01-04 9:54   3.75         NaN  United Kingdom  2010-01-04  
5    2010-01-04 9:54   0.85         NaN  United Kingdom  2010-01-04  
6   2010-01-04 10:28   2.55     14590.0  United Kingdom  2010-01-04  
7   2010-01-04 10:28   4.25     14590.0  United Kingdom  2010-01-04  
8   2010-01-04 10:28   2.55     14590.0  United Kingdom  2010-01-04  
9   2010-01-04 10:28   4.25     14590.0  United Kingdom  2010-01-04  
10  2010-01-04 10:28   6.95     14590.0  United Kingdom  2010-01-04  
11  2010-01-04 10:28   2.10     14590.0  United Kingdom  2010-01-04  
12  2010-01-04 10:33   7.95     14590.0  United Kingdom  2010-01-04  
13  2010-01-04 10:41   3.39     16550.0  United Kingdom  2010-01-04  
14  2010-01-04 10:43   5.95     13287.0  United Kingdom  2010-01-04  
15  2010-01-04 10:43   9.95     13287.0  United Kingdom  2010-01-04  
16  2010-01-04 10:43   4.95     13287.0  United Kingdom  2010-01-04  
17  2010-01-04 10:43   2.55     13287.0  United Kingdom  2010-01-04  
18  2010-01-04 10:43   2.25     13287.0  United Kingdom  2010-01-04  
19  2010-01-04 10:43   2.25     13287.0  United Kingdom  2010-01-04

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1011079 entries, 0 to 1022663
Data columns (total 9 columns):
 #   Column        Non-Null Count    Dtype  
---  ------        --------------    -----  
 0   Invoice       1011079 non-null  object 
 1   StockCode     1011079 non-null  object 
 2   Description   1006851 non-null  object 
 3   Quantity      1011079 non-null  int64  
 4   Invoice_time  1011079 non-null  object 
 5   Price         1011079 non-null  float64
 6   CustomerID    781578 non-null   float64
 7   Country       1011079 non-null  object 
 8   Purchase_dt   1011079 non-null  object 
dtypes: float64(2), int64(1), object(6)
memory usage: 77.1+ MB


In [11]:
df['Quantity'].describe()

count    1.011079e+06
mean     1.010526e+01
std      1.767742e+02
min     -8.099500e+04
25%      1.000000e+00
50%      3.000000e+00
75%      1.000000e+01
max      8.099500e+04
Name: Quantity, dtype: float64

In [12]:
df['Country'].value_counts()

United Kingdom          927374
EIRE                     17454
Germany                  17191
France                   13827
Netherlands               5041
Spain                     3727
Switzerland               3175
Belgium                   3091
Portugal                  2430
Australia                 1868
Channel Islands           1607
Italy                     1502
Sweden                    1359
Norway                    1287
Cyprus                    1125
Finland                   1049
Austria                    921
Denmark                    794
Unspecified                752
Greece                     622
Japan                      562
USA                        534
Poland                     505
United Arab Emirates       500
Israel                     368
Hong Kong                  360
Singapore                  346
Malta                      299
Canada                     228
Iceland                    222
RSA                        169
Lithuania                  154
Bahrain 

In [13]:
df['StockCode'].value_counts()

85123A    5496
22423     4306
85099B    4063
21212     3135
20725     3121
          ... 
79057B       1
21104        1
84448        1
20880        1
35630B       1
Name: StockCode, Length: 5288, dtype: int64

We found that：
    - Quantity has negative numbers
    - Description has null, there are test products in the Dataset
    - CustomerID has null
    - Most purchase are from UK

#### 3. Clean Data

- Delete Negative Quantity
- Delete null CustomerID
- Delete null Description
- Delete Description include 'a test product'
- Select Country with only United Kingdom

In [14]:
df = df.loc[df['Quantity'] > 0]

In [15]:
df['CustomerID'].isnull().sum()

225439

In [16]:
df[df['CustomerID'].isnull()].head()

Invoice StockCode                          Description  Quantity  \
3    493413     21724      PANDA AND BUNNIES STICKER SHEET         1   
4    493413     84578       ELEPHANT TOY WITH BLUE T-SHIRT         1   
5    493413     21723        ALPHABET HEARTS STICKER SHEET         1   
482  493485     72008  FROSTED BOX 9 WHITE T-LIGHT CANDLES        24   
483  493485     21589           SWALLOW GIANT TUBE MATCHES         2   

         Invoice_time  Price  CustomerID         Country Purchase_dt  
3     2010-01-04 9:54   0.85         NaN  United Kingdom  2010-01-04  
4     2010-01-04 9:54   3.75         NaN  United Kingdom  2010-01-04  
5     2010-01-04 9:54   0.85         NaN  United Kingdom  2010-01-04  
482  2010-01-04 14:48   0.42         NaN  United Kingdom  2010-01-04  
483  2010-01-04 14:48   2.55         NaN  United Kingdom  2010-01-04

Drop NA:

In [17]:
df = df.dropna(subset=['CustomerID'])

In [18]:
df = df.dropna(subset=['Description'])

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 763716 entries, 0 to 1022663
Data columns (total 9 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Invoice       763716 non-null  object 
 1   StockCode     763716 non-null  object 
 2   Description   763716 non-null  object 
 3   Quantity      763716 non-null  int64  
 4   Invoice_time  763716 non-null  object 
 5   Price         763716 non-null  float64
 6   CustomerID    763716 non-null  float64
 7   Country       763716 non-null  object 
 8   Purchase_dt   763716 non-null  object 
dtypes: float64(2), int64(1), object(6)
memory usage: 58.3+ MB


In [20]:
df['Description'].str.contains('a test product').sum()

12

In [21]:
df = df.loc[~df['Description'].str.contains('a test product')]

In [22]:
df = df.loc[df['Country'] == 'United Kingdom']

In [23]:
df['Country'].value_counts()

United Kingdom    685559
Name: Country, dtype: int64

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 685559 entries, 6 to 1022663
Data columns (total 9 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Invoice       685559 non-null  object 
 1   StockCode     685559 non-null  object 
 2   Description   685559 non-null  object 
 3   Quantity      685559 non-null  int64  
 4   Invoice_time  685559 non-null  object 
 5   Price         685559 non-null  float64
 6   CustomerID    685559 non-null  float64
 7   Country       685559 non-null  object 
 8   Purchase_dt   685559 non-null  object 
dtypes: float64(2), int64(1), object(6)
memory usage: 52.3+ MB


In [ ]:
df.to_csv('onlinepurchase_clean.csv', sep='|', index=False)

#### 4. Create user-to-item matrix

In [ ]:
customer_item_matrix = df.pivot_table(
    index='CustomerID', 
    columns='StockCode', 
    values='Quantity',
    aggfunc='sum'
)

In [ ]:
customer_item_matrix.head(20)

In [ ]:
customer_item_matrix = customer_item_matrix.applymap(lambda x: 1 if x > 0 else 0)

In [ ]:
customer_item_matrix.head(20)

#### 5. Create Collaborative filtering Porduct Recommendation

Use cosine_similarity in scikit-learn to calculate similarity

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

In [ ]:
user_user_sim_matrix = pd.DataFrame(cosine_similarity(customer_item_matrix))

In [ ]:
user_user_sim_matrix.head()

In [ ]:
user_user_sim_matrix.columns

In [ ]:
user_user_sim_matrix.index

In [ ]:
customer_item_matrix.index

In [ ]:
user_user_sim_matrix.columns = customer_item_matrix.index

In [ ]:
user_user_sim_matrix.head()

In [ ]:
user_user_sim_matrix.index = customer_item_matrix.index

In [ ]:
user_user_sim_matrix.head()

#### 6. Generate recommended product list

According to the last user-to-user matrix:

The similarity of the same user is 1
The similarity between user 12346 and user 12747 is 0.044721; the similarity between user 12748 and user 12346 is 0.058652.

Thus user 12748 is more similar to user 12346 than user 12747

In [ ]:
user_user_sim_matrix.head()

Take a single user (12346.0) as an example:

-First, sort the user's relevance to other users (including the user), from largest to smallest (using .loc[row index]

-Find the most similar other users, here is 17374.0

In [ ]:
user_user_sim_matrix.loc[12346.0].sort_values(ascending=False)

In [ ]:
customer_item_matrix.head()

In [ ]:
customer_item_matrix.loc[12346.0]

In [ ]:
customer_item_matrix.loc[12346.0].to_numpy().nonzero()

In [ ]:
len(customer_item_matrix.loc[12346.0].to_numpy().nonzero())

In [ ]:
customer_item_matrix.loc[12346.0].to_numpy().nonzero()[0]

In [ ]:
customer_item_matrix.loc[12346.0].iloc[customer_item_matrix.loc[12346.0].to_numpy().nonzero()[0]]

In [ ]:
customer_item_matrix.loc[12346.0].iloc[customer_item_matrix.loc[12346.0].to_numpy().nonzero()[0]].index

In [ ]:
items_bought_by_12346 = set(customer_item_matrix.loc[12346.0].iloc[
    customer_item_matrix.loc[12346.0].to_numpy().nonzero()[0]].index) 
print(items_bought_by_12346)

In [ ]:
items_bought_by_17374 = set(customer_item_matrix.loc[17374.0].iloc[
    customer_item_matrix.loc[17374.0].to_numpy().nonzero()[0]
].index)
items_bought_by_17374

-Before recommending, clarify who is recommended

-Here we use 12346.0 as the recommended target, and recommend user 17374.0 to buy items, but user 12346.0 has not purchased

-The final recommended product is obtained by subtracting the two purchase record sets obtained in the previous step

In [ ]:
items_to_recommend_to_12346 = items_bought_by_17374 - items_bought_by_12346
items_to_recommend_to_12346

-Finally, generate a description of the recommended product

In [ ]:
df.loc[df['StockCode'].isin(items_to_recommend_to_12346)][['StockCode', 'Description']]

In [ ]:
df.loc[df['StockCode'].isin(items_to_recommend_to_12346)][['StockCode', 'Description']].drop_duplicates()

#### 7. Streamline recommendation process

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

def get_user_user_sim_matrix(customer_item_matrix):
    
    """
    
    Arguments:
    customer_item_matrix(a Pandas DataFrame with customer as row index and item as column)
    
    Returns:
    user_user_sim_matrix, a Pandas DataFrame with customer as row index and column header, value is the cosine similarity)
    """
    
    user_user_sim_matrix = pd.DataFrame(cosine_similarity(customer_item_matrix))
    user_user_sim_matrix.columns = customer_item_matrix.index
    user_user_sim_matrix.index = customer_item_matrix.index
    
    return user_user_sim_matrix    

In [ ]:
customer_item_matrix.head()

In [ ]:
user_user_sim_matrix = get_user_user_sim_matrix(customer_item_matrix)

In [ ]:
user_user_sim_matrix.head()

In [ ]:
def get_recommendations(target_id, user_user_sim_matrix, customer_item_matrix):
    
    """
    
    Arguments:
    target_id, user_user_sim_matrix, customer_item_matrix
    
    Returns:
    recommend_df, a Pandas DataFrame with target_id, sim_userid, cos_value and recommend_prod
    """
        
    sim_rank = user_user_sim_matrix.loc[target_id].sort_values(ascending=False)
    
    sim_userid = sim_rank.index[1]
    cos_value = sim_rank.iloc[1]
    
    items_bought_by_target_id = set(customer_item_matrix.loc[target_id].iloc[
    customer_item_matrix.loc[target_id].nonzero()[0]].index) 
    
    items_bought_by_sim_userid = set(customer_item_matrix.loc[sim_userid].iloc[
    customer_item_matrix.loc[sim_userid].nonzero()[0]].index) 
    
    items_to_recommend_to_target_id = items_bought_by_sim_userid - items_bought_by_target_id
    
    recommend_df = pd.DataFrame(columns=['target_id', 'sim_userid', 'cos_value'], index=range(len(items_to_recommend_to_target_id)))
    
    recommend_df['target_id'] = target_id
    recommend_df['sim_userid'] = sim_userid
    recommend_df['cos_value'] = cos_value

    recommend_df['recommend_prod'] = list(items_to_recommend_to_target_id)
    
    return recommend_df

In [ ]:
get_recommendations(12746.0, user_user_sim_matrix, customer_item_matrix)